### ISSUE TITLE
Failure to rebuild some targets with shared dependency in parallel mode

### ISSUE THREAD

**Issue by mconwa01** [timbertson/gup#18]
When building multiple targets (b, c) that share a buildable dependency (a) in parallel mode (`-j n`), sometimes b or c is not updated when a is rebuilt. The following test fails consistently for me at the `yac` assertion:

```diff
diff --git a/test/test_parallel.py b/test/test_parallel.py
index b37534b..9a366e7 100644
--- a/test/test_parallel.py
+++ b/test/test_parallel.py
@@ -160,6 +160,24 @@ if not IS_WINDOWS:
 			self.build('--jobs=10', 'target', 'link/target')
 			self.assertEquals(self.read('target'), '1')
 
+		def test_multiple_targets_with_common_dependency(self):
+			self.write('input.txt', 'x')
+			self.write('a.gup', BASH + 'gup -u input.txt ; cat input.txt > "$1" ; echo -n a >> "$1"')
+			self.write('b.gup', BASH + 'gup -u a ; cat a > "$1" ; echo -n b >> "$1"')
+			self.write('c.gup', BASH + 'gup -u a ; cat a > "$1" ; echo -n c >> "$1"')
+			self.write('d.gup', BASH + 'gup -u c ; cat c > "$1" ; echo -n d >> "$1"')
+			# just building b and c can trigger the bug, but
+			# chaining c through d seems to be more reliable
+			self.build('--jobs=2', '-u', 'b', 'd')
+			self.assertEquals(self.read('b'), 'xab')
+			self.assertEquals(self.read('c'), 'xac')
+			self.assertEquals(self.read('d'), 'xacd')
+			self.write('input.txt', 'y')
+			self.build('--jobs=2', '-u', 'b', 'd')
+			self.assertEquals(self.read('b'), 'yab')
+			self.assertEquals(self.read('c'), 'yac')
+			self.assertEquals(self.read('d'), 'yacd')
+
 	class TestLocking(TestCase):
 		def test_deps_file_is_write_locked_during_build(self):
 			self.write('target.gup', re.sub(r'^\t{4}', '', '''
```

Other variations (such as `gup -u -j2 b c`) fail only some of the time, so it depends on the interleaving of parallel operations.

I think it may be related to python/gup/builder.py:53-5, where it checks whether a dependency has already been built in this invocation. The following change makes the above test pass (without breaking the others that I can run), but I don't know whether it causes any other issues:

```diff
diff --git a/python/bin/gup b/python/bin/gup
index 5ef5271..fb399ad 100755
--- a/python/bin/gup
+++ b/python/bin/gup
@@ -1612,7 +1612,7 @@ def _builder_build_if_dirty(target, allow_build):
 
 		if deps.already_built():
 			_builder_log.trace("CLEAN: %s has already been built in this invocation", target.path)
-			return False
+			return True
 
 		built_children = set()
 		def build_child_if_dirty(path):
diff --git a/python/gup/builder.py b/python/gup/builder.py
index e772517..1387126 100644
--- a/python/gup/builder.py
+++ b/python/gup/builder.py
@@ -52,7 +52,7 @@ def _build_if_dirty(target, allow_build):
 
 		if deps.already_built():
 			_log.trace("CLEAN: %s has already been built in this invocation", target.path)
-			return False
+			return True
 
 		built_children = set()
 		def build_child_if_dirty(path):
```

I wonder whether there's confusion here between "this target is clean, therefore we don't need to rebuild it" and "other targets that depend on it need to be rebuilt." For a target that has already been built in the current invocation, we don't need to build it again, but anything that depends on it does need to be rebuilt (if it hasn't already).

**Comment by timbertson** [comment]
Thanks for the detailed and reproducible report! ❤️ 

I figured this might take a bit to wrap my head around the ramifications, so apparently I procrastinated for 6+ months. The test case and fix looks good on first review.

I have some suspicions that the fix could potentially break other things though, my main suspicion is that this might cause checksum targets to behave incorrectly. So I'm taking a little more time to think that through.
